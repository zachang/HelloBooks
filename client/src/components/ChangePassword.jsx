import React from 'react';
import classnames from 'classnames';
import PropTypes from 'react-proptypes';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { changePasswordAction } from '../actions/userAction';

/**
 * ChangePassword class declaration
 *
 * @class ChangePassword
 *
 * @extends {React.Component}
 */
export class ChangePassword extends React.Component {
  /**
   * class constructor
   *
   * @param {object} props
   */
  constructor(props) {
    super(props);
    this.state = {
      changePasswordData: {
        oldPassword: '',
        newPassword: '',
        newPassword_confirmation: ''
      },
      errors: null,
      showToast: false,
    };
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  /**
   * @method componentWillReceiveProps
   *
   * @param {object} nextProps - nextProps
   *
   * @return {object} nextProps
   */
  componentWillReceiveProps(nextProps) {
    if (nextProps.userState.success === true) {
      if (this.state.showToast) {
        Materialize.toast('Password changed!', 4000);
        this.setState({
          showToast: false
        });
      }
    }
    if (nextProps.userState.errors) {
      this.setState({
        errors: nextProps.userState.errors
      });
    }
    if (!nextProps.userState.success &&
      nextProps.userState.fails === 'Incorrect old password') {
      if (this.state.showToast) {
        Materialize.toast('Incorrect old password!', 4000);
        this.setState({ showToast: false });
      }
      this.setState({
        errors: nextProps.userState.errors
      });
    }
    if (!nextProps.userState.success &&
      nextProps.userState.fails === 'Password not changed') {
      if (this.state.showToast) {
        Materialize.toast('Password not changed!', 4000);
        this.setState({ showToast: false });
      }
      this.setState({
        errors: nextProps.userState.errors
      });
    }
  }

  /**
   * Handles password update input
   *
   * @method handleChange
   *
   * @return {void} void
   *
   * @param {object} event - event
   */
  handleChange(event) {
    const changePasswordData = this.state.changePasswordData;
    changePasswordData[event.target.name] = event.target.value;
    this.setState({
      changePasswordData,
      errors: null
    });
  }

  /**
   * Handles password form update
   *
   * @method handleSubmit
   *
   * @return {void}
   *
   * @param {object} event - event
   */
  handleSubmit(event) {
    event.preventDefault();
    this.setState({
      changePasswordData: {
        oldPassword: '',
        newPassword: '',
        newPassword_confirmation: ''
      },
      showToast: true,
      imagePreviewUrl: '',
      errors: null
    });
    this.props.changePasswordAction(this.state.changePasswordData);
  }

  /**
   * Renders ChangePassword component
   *
   * @return {XML} JSX
   */
  render() {
    const { changePasswordData } = this.state;

    return (
      <div>
        {changePasswordData && <div>
          <div className='row'>
            <div className='col s10 m8 l6 bookadd'
              style={{ marginLeft: '35%', marginTop: '15%' }}
            >

              <div className='row'>
                <form className='col s10' onSubmit={this.handleSubmit}>
                  <div className='row'>
                    <div className='input-field col s10'>
                      <input
                        id='oldPassword'
                        name='oldPassword'
                        type='password'
                        className={classnames({
                          invalid: (this.state.errors &&
                            !!this.state.errors.oldPassword) ?
                            !!this.state.errors.oldPassword : false
                        })}
                        value={changePasswordData.oldPassword || ''}
                        onChange={this.handleChange}
                      />
                      <label
                        htmlFor='old password'
                        className={((this.state.errors &&
                          !!this.state.errors.oldPassword)
                          || changePasswordData.oldPassword.length > 0) ?
                          'custom-active custom-validate' : 'custom-validate'}
                        data-error={(this.state.errors &&
                          !!this.state.errors.oldPassword) ?
                          this.state.errors.oldPassword : ''}
                      >
                        Old Password
                      </label>
                    </div>
                  </div>
                  <div className='row'>
                    <div className='input-field col s10'>
                      <input
                        id='newPassword'
                        name='newPassword'
                        type='password'
                        className={classnames({
                          invalid: (this.state.errors &&
                            !!this.state.errors.newPassword) ?
                            !!this.state.errors.newPassword : false
                        })}
                        value={changePasswordData.newPassword}
                        onChange={this.handleChange}
                      />
                      <label
                        htmlFor='new password'
                        className={((this.state.errors &&
                          !!this.state.errors.newPassword)
                          || changePasswordData.newPassword.length > 0) ?
                          'custom-active custom-validate' : 'custom-validate'}
                        data-error={(this.state.errors &&
                          !!this.state.errors.newPassword) ?
                          this.state.errors.newPassword : ''}
                      >
                        New Password
                      </label>
                    </div>
                  </div>
                  <div className='row'>
                    <div className='input-field col s10'>
                      <input
                        id='newPassword_confirmation'
                        name='newPassword_confirmation'
                        type='password'
                        className={classnames({
                          invalid:
                            (this.state.errors &&
                              !!this.state.errors.newPassword_confirmation)
                              ? !!this.state.errors.newPassword_confirmation
                              : false
                        })}
                        value={changePasswordData.newPassword_confirmation}
                        onChange={this.handleChange}
                      />
                      <label
                        htmlFor='Confirm New Password'
                        className={
                          ((this.state.errors &&
                            !!this.state.errors.newPassword_confirmation) ||
                            changePasswordData.newPassword_confirmation.length >
                            0) ?
                            'custom-active custom-validate' : 'custom-validate'}
                        data-error={
                          (this.state.errors &&
                            !!this.state.errors.newPassword_confirmation) ?
                            this.state.errors.newPassword_confirmation : ''}
                      >
                        Confirm New Password
                      </label>
                    </div>
                  </div>

                  <button
                    className='col s10 btn btn-large waves-effect waves-light'
                    type='submit'
                    name='action'
                  >
                    Change Password
                  </button>

                </form>
              </div>
            </div>
          </div>
        </div>}
      </div>

    );
  }
}

ChangePassword.propTypes = {
  userState: PropTypes.object.isRequired,
  changePasswordAction: PropTypes.func.isRequired,
};
const mapStateToProps = state => ({
  userState: state.userReducer,
});
const mapDispatchToProps = dispatch =>
  bindActionCreators({ changePasswordAction }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(ChangePassword);