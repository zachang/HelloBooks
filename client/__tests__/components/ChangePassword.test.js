import React from 'react';
import toJson from 'enzyme-to-json';
import sinon from 'sinon';
import { shallow } from 'enzyme';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import storeConfiguration from '../../src/store/store';
import { ChangePassword } from '../../src/components/ChangePassword.jsx';

const event = {
  preventDefault: jest.fn(),
  target: {
    oldPassword: 'olpassword',
    newPassword: 'password',
    newPassword_confirmation: 'password'
  }
};

describe('<ChangePassword/>', () => {
  // create a spy function for signinAction
  const changePasswordAction = sinon.spy();
  // spy on handleSubmit of ChangePassword Page
  sinon.spy(ChangePassword.prototype, 'handleSubmit');
  // spy on handleChange of ChangePassword Page
  sinon.spy(ChangePassword.prototype, 'handleChange');
  // spy on componentWillReceiveProps of ChangePassword Page
  sinon.spy(ChangePassword.prototype, 'componentWillReceiveProps');

  const props = {
    changePasswordData: {
      oldPassword: '',
      newPassword: '',
      newPassword_confirmation: ''
    },
    userState: {
      success: false,
      errors: null,
      fails: null,
      users: [],
      user: [],
      pageCount: null,
    },
    changePasswordAction,
    showToast: false
  };

  const middleware = [thunk];
  const mockStore = configureMockStore(middleware);
  let store;
  let shallowComponent;
  beforeEach(() => {
    store = mockStore(storeConfiguration);
    shallowComponent = shallow(<ChangePassword {...props} />);
    shallowComponent.setState({
      showToast: true,
    });
    shallowComponent.setProps({
      userState: {
        success: true,
        errors: 'kkkkk',
        fails: '',
      }
    });
  });

  it('renders <ChangePassword /> component', () => {
    const tree = toJson(shallowComponent);
    expect(tree).toMatchSnapshot();
  });

  it('should check that handleChange is called', () => {
    shallowComponent.instance().handleChange(event);
    expect(ChangePassword.prototype.handleChange.calledOnce).toEqual(true);
  });

  it('should check that handleSubmit is called', () => {
    shallowComponent.instance().handleSubmit(event);
    expect(ChangePassword.prototype.handleSubmit.calledOnce).toEqual(true);
  });

  it('calls componentWillReceiveProps password is changed', () => {
    props.userState.sucess = true;
    shallowComponent.setState({
      showToast: true
    });
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(ChangePassword.prototype.componentWillReceiveProps.called).toEqual(true);
  });

  it('calls componentWillReceiveProps when an error occurs', () => {
    props.userState.errors = 'kkkkk';
    shallowComponent.setState({
      errors: props.userState.errors
    });
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(ChangePassword.prototype.componentWillReceiveProps.called).toEqual(true);
  });

  it('calls componentWillReceiveProps for Incorrect old password',
    () => {
      shallowComponent.setProps({
        userState: {
          success: false,
          errors: null,
          fails: 'Incorrect old password',
        }
      });
      shallowComponent.setState({
        errors: props.userState.errors
      });
      shallowComponent.instance().componentWillReceiveProps(props);
      expect(ChangePassword.prototype.componentWillReceiveProps
        .called).toEqual(true);
    });

  it('calls componentWillReceiveProps for Password not changed',
    () => {
      shallowComponent.setProps({
        userState: {
          success: false,
          errors: null,
          fails: 'Password not changed',
        }
      });
      shallowComponent.setState({
        errors: props.userState.errors
      });
      shallowComponent.instance().componentWillReceiveProps(props);
      expect(ChangePassword.prototype.componentWillReceiveProps
        .called).toEqual(true);
    });
});