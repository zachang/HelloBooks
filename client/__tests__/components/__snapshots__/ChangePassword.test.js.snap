// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChangePassword/> renders <ChangePassword /> component 1`] = `
<div>
  <div>
    <div
      className="row"
    >
      <div
        className="col s10 m8 l6 bookadd"
        style={
          Object {
            "marginLeft": "35%",
            "marginTop": "15%",
          }
        }
      >
        <div
          className="row"
        >
          <form
            className="col s10"
            onSubmit={[Function]}
          >
            <div
              className="row"
            >
              <div
                className="input-field col s10"
              >
                <input
                  className=""
                  id="oldPassword"
                  name="oldPassword"
                  onChange={[Function]}
                  type="password"
                  value=""
                />
                <label
                  className="custom-validate"
                  data-error=""
                  htmlFor="old password"
                >
                  Old Password
                </label>
              </div>
            </div>
            <div
              className="row"
            >
              <div
                className="input-field col s10"
              >
                <input
                  className=""
                  id="newPassword"
                  name="newPassword"
                  onChange={[Function]}
                  type="password"
                  value=""
                />
                <label
                  className="custom-validate"
                  data-error=""
                  htmlFor="new password"
                >
                  New Password
                </label>
              </div>
            </div>
            <div
              className="row"
            >
              <div
                className="input-field col s10"
              >
                <input
                  className=""
                  id="newPassword_confirmation"
                  name="newPassword_confirmation"
                  onChange={[Function]}
                  type="password"
                  value=""
                />
                <label
                  className="custom-validate"
                  data-error=""
                  htmlFor="Confirm New Password"
                >
                  Confirm New Password
                </label>
              </div>
            </div>
            <button
              className="col s10 btn btn-large waves-effect waves-light"
              name="action"
              type="submit"
            >
              Change Password
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
